File: c++/maxFlow2/main.cpp
#include <bits/stdc++.h>

using namespace std;
vector<long long> G[405];

vector<int> visited;

vector<int> dfs(int node) {
    visited.push_back(node);
    if (node == 400) {
        vector<int> p;
        p.push_back(400);
        return p;
    } else
        for (int neighbor : G[node]) {
            bool isContained = false;
            for (int elem : visited)
                if (elem == neighbor)
                    isContained = true;
            if (!isContained) {
                vector<int> neighborPath = dfs(neighbor);
                if (!neighborPath.empty()) {
                    neighborPath.insert(neighborPath.begin(), node);
                    return neighborPath;
                }
            }
        }
    return vector<int>();
}

int maxFlow() {
    visited.clear();
    vector<int> path = dfs(0);
    if (path.empty())
        return 0;
    for (int i = 0; i < path.size() - 1; i++) {
        int a = path[i];
        int b = path[i + 1];
        for (auto iter = G[a].begin(); iter != G[a].end(); iter++) {
            if (*iter == b) {
                G[a].erase(iter);
                break;
            }
        }
        G[b].push_back(a);
    }
    return maxFlow() + 1;
}

int main() {
    int n,m;
    cin >> n >> m;
    for(int i = 1 ; i <=n ; i++){
        int num;
        cin >> num;
        if(num == 0){
            G[0].push_back(i);
        }
        else{
            G[i].push_back(400);
        }
    }
    for(int i = 0 ; i<m ;i++){
        int v1,v2;
        cin >> v1 >> v2;
        G[v1].push_back(v2);
        G[v2].push_back(v1);
    }
    cout << maxFlow();
}

File: python/maxFlow2/main.py

def findMax():
        graph = {}
        nums = input().split(" ")
        n = int(nums[0])
        m = int(nums[1])
        for i in range(n):
                graph[i] = []
        graph[-1] = []
        graph[400] = []
        numbers = input().split(" ")
        a = 2
        for i in range(n):
                if numbers[i] == "0":
                        graph[-1] += [i]
                else:
                        graph[i] += [400]
        for j in range(m):
                inr = input().split(" ")
                v1 = int(inr[0]) -1
                v2 = int(inr[1]) -1
                if(len(graph[v1]) == 1 and len(graph[v2]) == 1):
                        graph[v1].append(v2)
                        graph[v2].append(v1)
                elif(len(graph[v1]) == 1 and len(graph[v2]) == 0):
                        graph[v2].append(v1)
                elif(len(graph[v1]) == 0 and len(graph[v2]) == 1):
                        graph[v1].append(v2)
                else:
                        graph[v1].append(v2)
                        graph[v2].append(v1)
        visited = []
        def dfs(graph,node):
                global visited
                visited.append(node)
                if node == 400:
                        return [400]
                for neighbor in graph[node]:
                        if neighbor not in visited:
                                visited.append(neighbor)
                                l = dfs(graph,neighbor)
                                if l != -1:
                                        return [node] + l
                return -1

        def maxFlow(graph):
                global visited
                visited = []
                l = dfs(graph,-1)

                if l == -1:
                        return 0
                for i in range(len(l) - 1):
                        a = l[i]
                        b = l[i + 1]
                        graph[a].remove(b)
                        graph[b].append(a)
                return maxFlow(graph) + 1
        return maxFlow(graph)

print(findMax())